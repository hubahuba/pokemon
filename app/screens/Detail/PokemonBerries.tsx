import * as React from 'react';
import {View} from 'react-native';
import Animated, {
  FadeInUp,
  FadeOutDown,
  interpolate,
} from 'react-native-reanimated';
import Carousel from 'react-native-reanimated-carousel';
import {GestureHandlerRootView} from 'react-native-gesture-handler';

import IconButton from '~atoms/IconButton';
import {PokemonBerriesProps} from '~/screens/Detail/Detail';
import {memo, useState} from 'react';
import BerryButton from '~atoms/BerryButton';

const PokemonBerries = memo(function PokemonBerries({
  data,
  onCloseBerry,
  onSelectBerry,
  ownedId,
}: PokemonBerriesProps) {
  const [viewWidth, setViewWidth] = useState(0);
  const itemSize = 80;
  const iconSize = 34;
  const centerOffset = viewWidth / 2 - itemSize / 2;

  const animationStyle = React.useCallback(
    (value: number) => {
      'worklet';

      const itemGap = interpolate(
        value,
        [-3, -2, -1, 0, 1, 2, 3],
        [-30, -10, 0, 0, 0, 10, 30],
      );

      const translateX =
        interpolate(value, [-1, 0, 1], [-itemSize, 0, itemSize]) +
        centerOffset -
        itemGap;

      const translateY = interpolate(
        value,
        [-1, -0.3, 0, 0.4, 1],
        [50, 35, 30, 35, 50],
      );

      const scale = interpolate(
        value,
        [-1, -0.5, 0, 0.5, 1],
        [0.8, 0.85, 1.1, 0.85, 0.8],
      );

      return {
        transform: [
          {
            translateX,
          },
          {
            translateY,
          },
          {scale},
        ],
      };
    },
    [centerOffset],
  );

  if (!ownedId) {
    return <View />;
  }

  return (
    <Animated.View
      testID="PokemonBerriesWrapper"
      onLayout={event => setViewWidth(event.nativeEvent.layout.width)}
      entering={FadeInUp}
      exiting={FadeOutDown}>
      <GestureHandlerRootView>
        <Carousel
          width={itemSize}
          height={itemSize}
          style={{
            width: '100%',
            height: viewWidth / 2,
          }}
          loop
          autoPlay={false}
          data={data}
          renderItem={({item}) => (
            <BerryButton
              item={item}
              onPress={onSelectBerry}
              ownedId={ownedId}
              testID="SelectBerry"
            />
          )}
          customAnimation={animationStyle}
        />
        <IconButton
          testID="CloseBerriesButton"
          className="absolute rounded-full bg-gray-400 p-1 border border-gray-500"
          style={{
            top: viewWidth / 2 - iconSize,
            left: viewWidth / 2 - iconSize / 2,
          }}
          iconName="close"
          onPress={onCloseBerry}
        />
      </GestureHandlerRootView>
    </Animated.View>
  );
});

export default PokemonBerries;
