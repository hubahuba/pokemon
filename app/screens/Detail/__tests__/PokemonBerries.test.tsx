import React from 'react';
import PokemonBerries from '~/screens/Detail/PokemonBerries';
import {dataBerry} from '@/__mocks__/constanta.ts';
import {
  fireEvent,
  render,
  screen,
  waitFor,
  waitForElementToBeRemoved,
} from '@testing-library/react-native';

const mockClose = jest.fn();
const mockSelectBerry = jest.fn();

describe('PokemonBerries', () => {
  beforeEach(async () => {
    render(
      <PokemonBerries
        onCloseBerry={mockClose}
        onSelectBerry={mockSelectBerry}
        data={dataBerry}
        ownedId="randomString"
      />,
    );
    await waitForElementToBeRemoved(() => {
      return screen.getByTestId('__CAROUSEL_ITEM_1_NOT_READY__');
    });
  });

  it('renders PokemonBerries correctly', async () => {
    const tree = screen.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders PokemonBerries with no ownedId correctly', () => {
    const tree = render(<PokemonBerries data={dataBerry} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('PokemonBerries close action test', async () => {
    const button = screen.getByTestId('CloseBerriesButton');
    fireEvent.press(button);
    await waitFor(() => expect(mockClose.mock.calls.length).toBe(1));
  });

  test('PokemonBerries select action test', async () => {
    const button = screen.getAllByTestId('SelectBerry');
    fireEvent.press(button[0]);
    await waitFor(() => {
      expect(mockSelectBerry.mock.calls.length).toBe(1);
    });
  });
});
