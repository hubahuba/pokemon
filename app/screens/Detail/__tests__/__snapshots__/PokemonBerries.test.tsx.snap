// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonBerries renders PokemonBerries correctly 1`] = `
<View
  collapsable={false}
  entering={[Function]}
  exiting={[Function]}
  jestAnimatedStyle={
    {
      "value": {},
    }
  }
  onLayout={[Function]}
  testID="PokemonBerriesWrapper"
>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      style={
        {
          "flex": 1,
        }
      }
    >
      <View
        collapsable={false}
        forwardedRef={[Function]}
        handlerTag={1}
        handlerType="PanGestureHandler"
        jestAnimatedStyle={
          {
            "value": {},
          }
        }
        onGestureHandlerEvent={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          {
            "flexDirection": "row",
            "height": 0,
            "overflow": "hidden",
            "width": "100%",
          }
        }
      >
        <View
          collapsable={false}
          jestAnimatedStyle={
            {
              "value": {
                "transform": [
                  {
                    "translateX": -40,
                  },
                  {
                    "translateY": 30,
                  },
                  {
                    "scale": 1.1,
                  },
                ],
              },
            }
          }
          style={
            {
              "height": 80,
              "position": "absolute",
              "transform": [
                {
                  "translateX": -40,
                },
                {
                  "translateY": 30,
                },
                {
                  "scale": 1.1,
                },
              ],
              "width": 80,
            }
          }
          testID="__CAROUSEL_ITEM_0_READY__"
        >
          <RNGestureHandlerButton
            collapsable={false}
            delayLongPress={600}
            enabled={true}
            exclusive={true}
            handlerTag={2}
            handlerType="NativeViewGestureHandler"
            onGestureEvent={[Function]}
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            onHandlerStateChange={[Function]}
            rippleColor={0}
            testID="SelectBerry"
            touchSoundDisabled={false}
          >
            <View
              accessible={true}
              collapsable={false}
              style={
                {
                  "height": "100%",
                  "opacity": 1,
                  "width": "100%",
                }
              }
            >
              <View
                style={
                  [
                    {
                      "display": "flex",
                    },
                    {
                      "flexBasis": "0%",
                      "flexGrow": 1,
                      "flexShrink": 1,
                    },
                    {
                      "paddingBottom": 2,
                      "paddingLeft": 2,
                      "paddingRight": 2,
                      "paddingTop": 2,
                    },
                    {
                      "borderBottomLeftRadius": 9999,
                      "borderBottomRightRadius": 9999,
                      "borderTopLeftRadius": 9999,
                      "borderTopRightRadius": 9999,
                    },
                    {
                      "backgroundColor": "#f3f4f6",
                    },
                    {
                      "justifyContent": "center",
                    },
                    {
                      "alignItems": "center",
                    },
                    {
                      "overflow": "hidden",
                    },
                  ]
                }
              >
                <View
                  style={
                    [
                      {
                        "overflow": "hidden",
                      },
                      [
                        {
                          "width": "100%",
                        },
                        {
                          "height": "100%",
                        },
                      ],
                    ]
                  }
                >
                  <FastImageView
                    defaultSource={
                      {
                        "testUri": "../../../app/assets/images/berry.png",
                      }
                    }
                    resizeMode="cover"
                    style={
                      {
                        "bottom": 0,
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                      }
                    }
                  />
                </View>
              </View>
            </View>
          </RNGestureHandlerButton>
        </View>
        <View
          collapsable={false}
          jestAnimatedStyle={
            {
              "value": {
                "transform": [
                  {
                    "translateX": 40,
                  },
                  {
                    "translateY": 50,
                  },
                  {
                    "scale": 0.8,
                  },
                ],
              },
            }
          }
          style={
            {
              "height": 80,
              "position": "absolute",
              "transform": [
                {
                  "translateX": 40,
                },
                {
                  "translateY": 50,
                },
                {
                  "scale": 0.8,
                },
              ],
              "width": 80,
            }
          }
          testID="__CAROUSEL_ITEM_1_READY__"
        >
          <RNGestureHandlerButton
            collapsable={false}
            delayLongPress={600}
            enabled={true}
            exclusive={true}
            handlerTag={3}
            handlerType="NativeViewGestureHandler"
            onGestureEvent={[Function]}
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            onHandlerStateChange={[Function]}
            rippleColor={0}
            testID="SelectBerry"
            touchSoundDisabled={false}
          >
            <View
              accessible={true}
              collapsable={false}
              style={
                {
                  "height": "100%",
                  "opacity": 1,
                  "width": "100%",
                }
              }
            >
              <View
                style={
                  [
                    {
                      "display": "flex",
                    },
                    {
                      "flexBasis": "0%",
                      "flexGrow": 1,
                      "flexShrink": 1,
                    },
                    {
                      "paddingBottom": 2,
                      "paddingLeft": 2,
                      "paddingRight": 2,
                      "paddingTop": 2,
                    },
                    {
                      "borderBottomLeftRadius": 9999,
                      "borderBottomRightRadius": 9999,
                      "borderTopLeftRadius": 9999,
                      "borderTopRightRadius": 9999,
                    },
                    {
                      "backgroundColor": "#f3f4f6",
                    },
                    {
                      "justifyContent": "center",
                    },
                    {
                      "alignItems": "center",
                    },
                    {
                      "overflow": "hidden",
                    },
                  ]
                }
              >
                <View
                  style={
                    [
                      {
                        "overflow": "hidden",
                      },
                      [
                        {
                          "width": "100%",
                        },
                        {
                          "height": "100%",
                        },
                      ],
                    ]
                  }
                >
                  <FastImageView
                    defaultSource={
                      {
                        "testUri": "../../../app/assets/images/berry.png",
                      }
                    }
                    resizeMode="cover"
                    style={
                      {
                        "bottom": 0,
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                      }
                    }
                  />
                </View>
              </View>
            </View>
          </RNGestureHandlerButton>
        </View>
        <View
          collapsable={false}
          jestAnimatedStyle={
            {
              "value": {
                "transform": [
                  {
                    "translateX": 110,
                  },
                  {
                    "translateY": 75,
                  },
                  {
                    "scale": 0.7000000000000002,
                  },
                ],
              },
            }
          }
          style={
            {
              "height": 80,
              "position": "absolute",
              "transform": [
                {
                  "translateX": 110,
                },
                {
                  "translateY": 75,
                },
                {
                  "scale": 0.7000000000000002,
                },
              ],
              "width": 80,
            }
          }
          testID="__CAROUSEL_ITEM_2_READY__"
        >
          <RNGestureHandlerButton
            collapsable={false}
            delayLongPress={600}
            enabled={true}
            exclusive={true}
            handlerTag={4}
            handlerType="NativeViewGestureHandler"
            onGestureEvent={[Function]}
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            onHandlerStateChange={[Function]}
            rippleColor={0}
            testID="SelectBerry"
            touchSoundDisabled={false}
          >
            <View
              accessible={true}
              collapsable={false}
              style={
                {
                  "height": "100%",
                  "opacity": 1,
                  "width": "100%",
                }
              }
            >
              <View
                style={
                  [
                    {
                      "display": "flex",
                    },
                    {
                      "flexBasis": "0%",
                      "flexGrow": 1,
                      "flexShrink": 1,
                    },
                    {
                      "paddingBottom": 2,
                      "paddingLeft": 2,
                      "paddingRight": 2,
                      "paddingTop": 2,
                    },
                    {
                      "borderBottomLeftRadius": 9999,
                      "borderBottomRightRadius": 9999,
                      "borderTopLeftRadius": 9999,
                      "borderTopRightRadius": 9999,
                    },
                    {
                      "backgroundColor": "#f3f4f6",
                    },
                    {
                      "justifyContent": "center",
                    },
                    {
                      "alignItems": "center",
                    },
                    {
                      "overflow": "hidden",
                    },
                  ]
                }
              >
                <View
                  style={
                    [
                      {
                        "overflow": "hidden",
                      },
                      [
                        {
                          "width": "100%",
                        },
                        {
                          "height": "100%",
                        },
                      ],
                    ]
                  }
                >
                  <FastImageView
                    defaultSource={
                      {
                        "testUri": "../../../app/assets/images/berry.png",
                      }
                    }
                    resizeMode="cover"
                    style={
                      {
                        "bottom": 0,
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                      }
                    }
                  />
                </View>
              </View>
            </View>
          </RNGestureHandlerButton>
        </View>
        <View
          collapsable={false}
          jestAnimatedStyle={
            {
              "value": {
                "transform": [
                  {
                    "translateX": -120,
                  },
                  {
                    "translateY": 50,
                  },
                  {
                    "scale": 0.8,
                  },
                ],
              },
            }
          }
          style={
            {
              "height": 80,
              "position": "absolute",
              "transform": [
                {
                  "translateX": -120,
                },
                {
                  "translateY": 50,
                },
                {
                  "scale": 0.8,
                },
              ],
              "width": 80,
            }
          }
          testID="__CAROUSEL_ITEM_3_READY__"
        >
          <RNGestureHandlerButton
            collapsable={false}
            delayLongPress={600}
            enabled={true}
            exclusive={true}
            handlerTag={5}
            handlerType="NativeViewGestureHandler"
            onGestureEvent={[Function]}
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            onHandlerStateChange={[Function]}
            rippleColor={0}
            testID="SelectBerry"
            touchSoundDisabled={false}
          >
            <View
              accessible={true}
              collapsable={false}
              style={
                {
                  "height": "100%",
                  "opacity": 1,
                  "width": "100%",
                }
              }
            >
              <View
                style={
                  [
                    {
                      "display": "flex",
                    },
                    {
                      "flexBasis": "0%",
                      "flexGrow": 1,
                      "flexShrink": 1,
                    },
                    {
                      "paddingBottom": 2,
                      "paddingLeft": 2,
                      "paddingRight": 2,
                      "paddingTop": 2,
                    },
                    {
                      "borderBottomLeftRadius": 9999,
                      "borderBottomRightRadius": 9999,
                      "borderTopLeftRadius": 9999,
                      "borderTopRightRadius": 9999,
                    },
                    {
                      "backgroundColor": "#f3f4f6",
                    },
                    {
                      "justifyContent": "center",
                    },
                    {
                      "alignItems": "center",
                    },
                    {
                      "overflow": "hidden",
                    },
                  ]
                }
              >
                <View
                  style={
                    [
                      {
                        "overflow": "hidden",
                      },
                      [
                        {
                          "width": "100%",
                        },
                        {
                          "height": "100%",
                        },
                      ],
                    ]
                  }
                >
                  <FastImageView
                    defaultSource={
                      {
                        "testUri": "../../../app/assets/images/berry.png",
                      }
                    }
                    resizeMode="cover"
                    style={
                      {
                        "bottom": 0,
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                      }
                    }
                  />
                </View>
              </View>
            </View>
          </RNGestureHandlerButton>
        </View>
      </View>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#9ca3af",
          "borderBottomColor": "#6b7280",
          "borderBottomLeftRadius": 9999,
          "borderBottomRightRadius": 9999,
          "borderBottomWidth": 1,
          "borderLeftColor": "#6b7280",
          "borderLeftWidth": 1,
          "borderRightColor": "#6b7280",
          "borderRightWidth": 1,
          "borderTopColor": "#6b7280",
          "borderTopLeftRadius": 9999,
          "borderTopRightRadius": 9999,
          "borderTopWidth": 1,
          "left": -17,
          "opacity": 1,
          "paddingBottom": 4,
          "paddingLeft": 4,
          "paddingRight": 4,
          "paddingTop": 4,
          "position": "absolute",
          "top": -34,
        }
      }
      testID="CloseBerriesButton"
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": undefined,
              "fontSize": 24,
            },
            undefined,
            {
              "fontFamily": "anticon",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
</View>
`;

exports[`PokemonBerries renders PokemonBerries with no ownedId correctly 1`] = `<View />`;
